crates/proto/src/udp/udp_stream.rs

```diff
@@ -273,22 +284,31 @@ impl<P: RuntimeProvider> Future for NextRandomUdpSocket<P> {
                     }
                 },
                 None => {
-                    let bind_addr = match this.bind_address.port() {
-                        0 if this.attempted < ATTEMPT_RANDOM => {
+                    let mut bind_addr = this.bind_address;
+                    if bind_addr.port() == 0 {
+                        while this.attempted < ATTEMPT_RANDOM {
                             // Per RFC 6056 Section 3.2:
                             //
                             // As mentioned in Section 2.1, the dynamic ports consist of the range
                             // 49152-65535.  However, ephemeral port selection algorithms should use
                             // the whole range 1024-65535.
                             let rand_port_range = Uniform::new_inclusive(1_024, u16::MAX);
                             let mut rand = rand::thread_rng();
-                            SocketAddr::new(
-                                this.bind_address.ip(),
-                                rand_port_range.sample(&mut rand),
-                            )
+                            let port = rand_port_range.sample(&mut rand);
+                            if this
+                                .avoid_local_ports
+                                .as_ref()
+                                .map(|set| set.contains(&port))
+                                .unwrap_or_default()
+                            {
+                                this.attempted += 1;
```

Comment:
I think this is correct but it is a little nuanced because it's easy to confuse "attempts at picking a port" with "attempts to bind", I suppose? Maybe add a comment about why this is correct?